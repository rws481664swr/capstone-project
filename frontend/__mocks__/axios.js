import {BASE_URL} from "../src/config";

axios.get.mockImplementation((url) => {
    let result = url.slice(BASE_URL.length);
    if(result[result.length-1] === '/') result = result.slice(0, result.length-1);
    switch (result){
        case '/courses':
            return Promise.resolve({data: coursesList})
        case '/users':
            return Promise.resolve({data:usersList})
        case '/auth/token':
            return Promise.resolve({data:'token'})
    }
    return Promise.resolve({data: []})

})
const axios = {
    get: jest.fn().mockResolvedValue({
        data: {}
    }),
    post: jest.fn().mockResolvedValue({
        data: {}
    }),
    put: jest.fn().mockResolvedValue({
        data: {}
    }),
    delete: jest.fn().mockResolvedValue({
        data: {}
    })
}
export default axios