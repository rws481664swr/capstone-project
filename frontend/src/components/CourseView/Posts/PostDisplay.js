import useAxios from "../../../api";
import useToggle from "../../../hooks/useToggle";
import Modal, {useModal} from "../../General/Modal";
import React, {useCallback} from "react";
import Comments from "./Comments";
import {useGlobalContext} from "../../../state/contexts/GlobalContext";
import {EditButton, PinButton} from "./EditPost/Buttons";
import Edit from "./EditPost/Edit";
import {useDispatch} from "react-redux";
import {Link} from "react-router-dom";
import './PostDisplay.css'
import dayjs from "dayjs";
const useDisplayPost = ([post, setPost]) => {
    const axios = useAxios()
    const [visible, {hide: hideModal}] = useModal()
    const dispatch = useDispatch()
    const hide = useCallback(() => {
        setPost(null)
        hideModal()
    }, [setPost, hideModal])
    const togglePin = useCallback(async () => {
        try {
            const payload = await axios.put(`posts/${post._id}/pin`)
            dispatch({type: payload.pinned ? "PIN" : "UNPIN", payload})
            setPost(payload)
        } catch (e) {
            alert(e.response.data.message)
            console.error(e.response.data.message)
        }
    }, [dispatch, post, setPost, axios])
    return {togglePin, hide, visible}
}
const PostDisplay = ({post = null, setPost}) => {
    const {togglePin, hide:hideModal} = useDisplayPost([post, setPost])
    const [editMode, toggleEditMode,setEditMode] = useToggle(false)
    const hide = ()=> {
        hideModal()
        setEditMode(false)
    }
    const {role, username: currentUser} = useGlobalContext()
    if (!post) return null
    const {
        username,
        title,
        content,
        postDate,
        _id,
        pinned,
        comments,
        user,
        course
    } = post
    let date = dayjs(new Date(postDate))
    const part1_date = `${date.format(' h:mm A')}`
    const part2_date = ` ${date.format('M/DD')}`
    date = part1_date + part2_date
    const editable = currentUser === username
    const canPin = role !== "STUDENT"
    return (
        <Modal className={'PostDisplay'} visible={post} hide={hide}>
            <div className={'split-between'}>

                {/*<PostedByOn postDate={postDate} username={username}/>*/}
            </div>

           <div className={'PostDisplayHeading'}>
               <h4 className={'PostDisplay-Title'}>{title}</h4>
               <div className="PostDisplay-Buttons">
                   <EditButton className={'EditButton'}
                               editable={editable}
                               editMode={editMode}
                               setEditMode={toggleEditMode}/>
                   <PinButton className={'PinButton'}
                              canPin={canPin}
                              pinned={pinned}
                              togglePin={togglePin}/>
               </div>
        </div>
            <span className={'PostedByOn'}>  posted by
                <Link to={`/users/${username}`}>  {username}  </Link>  at
                <span className={'PostDisplay-TimeStamp'}>{date}</span>
            </span>
            <div>{pinned && 'pinned'}</div>

            <p className={'PostDisplay-Content'}> {content}</p>

            <div test-data={_id}></div>
            {editMode && <Edit editMode={editMode}
                               post={post}
                               setPost={setPost}
                               setEditMode={toggleEditMode}/>}
            <Comments post={post}/>

        </Modal>

    )
}


export default PostDisplay
