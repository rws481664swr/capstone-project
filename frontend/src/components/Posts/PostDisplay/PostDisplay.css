
div.PostDisplay {
    /*display: flex;*/
    list-style: none;
    flex-direction: column;
    justify-content: space-between;
    align-items: flex-start;
    padding: 1rem;
    border: 1px solid #ccc;
    border-radius: 5px;
    margin-bottom: 1rem;
    min-height: 10rem;
    font-size: 1.6rem;
    color: #333;
    text-decoration: none;
    transition: all 0.3s ease-out;
}

.split-between {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}




.PostedByOn {

    font-size: 1rem;
    font-weight: 300;
    border: none;

}

.PostedByOn a span{
    cursor: pointer;
}

.PostDisplayHeading {
    display: flex;
    justify-content: space-between;
}


.PostDisplay-Title {
    font-size: 1.6rem;
    font-weight: 600;
    color: #000;
    text-decoration: none;
    cursor: pointer;
    background-color: #fff;
    border: none;
}

.PostDisplay-Content {
    font-size: 1rem;
    color: #333;
    text-decoration: none;

}
.CourseFormFlash{
    font-size: 1.4rem;
    font-weight: 600;
    color: #000;
    text-decoration: none;
    cursor: pointer;
    background-color: #fff;
    border: none;
}
.PostDisplay-ShowComments {
    cursor: pointer;

    text-decoration: none;
    color: #000;
    font-size: 1rem;
    font-weight: 300;
    background-color: #fff;
    border: none;
}

