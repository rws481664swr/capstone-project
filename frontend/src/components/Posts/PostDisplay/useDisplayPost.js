import useAxios from "../../../hooks/ajax/useAxios";
import {useDispatch} from "react-redux";
import {useMemo} from "react";
import {REMOVE} from "../../../state/actions/actions";

const useDisplayPost = ([post, setPost], setEditMode) => {
    const axios = useAxios()
    const dispatch = useDispatch()
    return useMemo(
        () => ({
            togglePin: async () => {
                try {
                    await axios.put(`posts/${post._id}/pin`)
                    const payload = {...post, pinned: !post.pinned}
                    dispatch({type: payload.pinned ? "PIN" : "UNPIN", payload: payload})
                    setPost(payload,`setPost called in memo`)
                } catch (e) {
                    console.error(e.response.data.message)
                }
            },

            deletePost: async () => {
                try {
                    const payload = await axios.delete(`posts/${post._id}`)
                    dispatch({type: REMOVE, id: post._id})
                    setPost(null,`setPost called in memo`)
                } catch (e) {
                    console.error(e.response.data.message)
                }
            }
        }), [post ,setPost, dispatch, axios, setEditMode])
}

export default useDisplayPost