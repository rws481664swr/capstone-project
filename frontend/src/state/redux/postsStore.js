import {configureStore} from "@reduxjs/toolkit";
import postsReducer from "./postsReducer";

const postsStore =
    configureStore({
        reducer: postsReducer
    })
export default postsStore