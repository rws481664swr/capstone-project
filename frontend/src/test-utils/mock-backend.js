import {BASE_URL} from "../config";

export const extractKeysAsList= (obj)=>
    Object.keys(obj).map(key=>obj[key])
        .reduce((acc, val)=>[...acc,val],[])
export const MOCK_TOKEN = {current: 'TOKEN'}

export const spliceUrl = (url) => {
    const arr = BASE_URL.split('')
    const toReturn = url.split('')
    while (arr.length){
        arr.shift()
        toReturn.shift()
    }
    return toReturn.join('')

}