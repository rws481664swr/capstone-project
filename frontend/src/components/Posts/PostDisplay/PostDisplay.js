import './PostDisplay.css'
import useToggle from "../../../hooks/state/useToggle";
import Modal from "../../General/Modal/Modal";
import React from "react";
import Comments from "../../Comments";
import {useGlobalContext} from "../../../state/contexts/GlobalContext";
import {DeleteButton, EditButton, PinButton} from "../../General/Button/IconButtons/IconButtons";
import Edit from "../EditPost/Edit";
import {Link} from "react-router-dom";
import Timestamp from "../../General/Timestamp";
import useDisplayPost from "./useDisplayPost";


const PostDisplay = ({post = null, setPost}) => {
    const [editMode, toggleEditMode, setEditMode] = useToggle(false)
    const {togglePin, hide, deletePost} = useDisplayPost([post, setPost],setEditMode)

    console.log(`rendering post display`)
    const PostButtons = ({editable,editMode, canPin, pinned, togglePin, toggleEditMode, deletePost}) =>
        <div className="PostDisplay-Buttons">

            <PinButton className={'PinButton'}
                       canPin={canPin}
                       pinned={pinned}
                       togglePin={togglePin}/>
            {editable && <>
                <EditButton className={'EditButton'}
                            editable={editable}
                            editMode={editMode}
                            setEditMode={toggleEditMode}/>

            </>}
            {editMode &&
                <DeleteButton className={`DeleteButton`}
                              onClick={deletePost
                                  // () => setDeleteConfirmation(true)
                              }/>
            }

        </div>
    const {role, username: currentUser} = useGlobalContext()
    if (!post) return null
    const {
        username,
        title,
        _id,
        pinned,
        content,
        postDate
    } = post
    console.log(`Post id: ${_id}`)
    const editable = currentUser === username
    const canPin = role !== "STUDENT"
    return (

        <Modal className={'PostDisplay'} visible={post} hide={hide}>
            <div className={'split-between'}></div>
            <div className={'PostDisplayHeading'}>
                <h4 className={'PostDisplay-Title'}>{title}</h4>
                <PostButtons
                    editable={editable}
                    canPin={canPin}
                    pinned={pinned}
                    togglePin={togglePin}
                    toggleEditMode={toggleEditMode}
                    deletePost={deletePost}
                    editMode={editMode}
                />
            </div>

            <PostedByOn username={username} postDate={postDate}/>
            <hr/>
            <p className={'PostDisplay-Content'}> {content}</p>

           <Edit editMode={editMode}
                               post={post}
                               setPost={setPost}
                               setEditMode={toggleEditMode}/>
            {post && !editMode && <Comments post={post}/>}

        </Modal>

    )
}

const PostedByOn = ({username, postDate}) =>
    <span className={'PostedByOn'}>
        <span> posted by </span>
           <span>
               <Link to={`/users/${username}`}>
                    {username}
                </Link>
           </span>
            <span className={' PostDisplay-TimeStamp'}>
                <Timestamp date={postDate}/>
            </span>
    </span>


export default PostDisplay
