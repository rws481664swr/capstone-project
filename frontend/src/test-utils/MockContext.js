import GlobalContext from "../state/contexts/GlobalContext";
import {MOCK_TOKEN} from "./mock-backend";
import {MemoryRouter} from "react-router-dom";

const MockContext = ({children, isAdmin = false, ...mockValue}) => {
    return (
        <MemoryRouter>
            <GlobalContext.Provider value={{
                token: MOCK_TOKEN,
                _id: 'id',
                username: 'username',
                role: isAdmin ? "ADMIN" : "STUDENT",
                loggedIn: true,
                timestamp: new Date(),
                ...mockValue
            }}>
                {children}
            </GlobalContext.Provider>
        </MemoryRouter>
    )
}

export default MockContext