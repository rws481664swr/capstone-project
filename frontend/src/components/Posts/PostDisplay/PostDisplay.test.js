import {render,fireEvent} from '@testing-library/react'
import PostDisplay from "./PostDisplay";
import MockContext from "../../../test-utils/MockContext";
import {Provider} from "react-redux";
import postsStore from "../../../state/redux/postsStore";
import {act} from "react-dom/test-utils";
import axios from "axios";
import getData from "../../../mock_data/data";
import {MOCK_TOKEN} from "../../../../mock-backend";

let post, comment
beforeEach(() => {
    const {postsList} = getData()
    post = postsList[0]
    comment =
        {
            "post": post._id,
            "username": "user",
            "content": "Culpa nisi delectus voluptate.",
            "timestamp": "2023-05-24T15:56:23.258Z",
            "_id": "646e33a759eab5cd39bbb8d3",
            "__v": 0
        }
})
jest.mock('axios', () => ({
    get: jest.fn(),
    post: jest.fn(),
    put: jest.fn(),
    delete: jest.fn()
}))
const axiosMock =
    async (url,body,config) => {
        if (url.includes('comments')) {
            return {data: [comment]}
        } else if (url.includes('posts')) {
            return {data: [post]}
        } else if (url.includes('auth/token')) {
            return {data: {token: MOCK_TOKEN.current}}
        }
        throw new Error()
    }
it('should render PostDisplay', function () {
axios.get.mockImplementation(axiosMock)
    const val = {post:null }
    const {asFragment} = render(
        <MockContext>
            <Provider store={postsStore}>
            <PostDisplay post={val.post} setPost={(v)=>val.post=v} />
            </Provider>
        </MockContext>)
    expect(asFragment()).toMatchSnapshot()
});it('should render PostDisplay with post', function () {
axios.get.mockImplementation(axiosMock)
    const val = {post}
    const {asFragment} = render(
        <MockContext>
            <Provider store={postsStore}>
            <PostDisplay post={val.post} setPost={(v)=>val.post=v} />
            </Provider>
        </MockContext>)
    expect(asFragment()).toMatchSnapshot()
});