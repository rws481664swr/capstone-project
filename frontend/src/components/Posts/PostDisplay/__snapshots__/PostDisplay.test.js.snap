// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostDisplay should render PostDisplay Comments with post 1`] = `
<DocumentFragment>
  <div
    class="sr-modal"
    style="z-index: 1;"
  >
    <div
      class="PostDisplay sr-modal_content"
      style="z-index: 1;"
    >
      <div
        class="split-between"
      />
      <div
        class="PostDisplayHeading"
        data-testid="PostDisplay-Heading"
      >
        <h4
          class="PostDisplay-Title"
        >
          Nulla saepe et nemo molestiae ducimus provident.
        </h4>
        <div
          class="PostDisplay-Buttons"
        >
          <button
            class="PinButton  PinButton IconButton  "
            data-testid="PinButton"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-thumbtack "
              data-icon="thumbtack"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 384 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M32 32C32 14.3 46.3 0 64 0H320c17.7 0 32 14.3 32 32s-14.3 32-32 32H290.5l11.4 148.2c36.7 19.9 65.7 53.2 79.5 94.7l1 3c3.3 9.8 1.6 20.5-4.4 28.8s-15.7 13.3-26 13.3H32c-10.3 0-19.9-4.9-26-13.3s-7.7-19.1-4.4-28.8l1-3c13.8-41.5 42.8-74.8 79.5-94.7L93.5 64H64C46.3 64 32 49.7 32 32zM160 384h64v96c0 17.7-14.3 32-32 32s-32-14.3-32-32V384z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
      <span
        class="PostedByOn"
      >
        <span>
           posted by 
        </span>
        <span>
          <a
            href="/users/user"
          >
            user
          </a>
        </span>
        <span
          class=" PostDisplay-TimeStamp"
        >
          <span
            class=""
          >
              at  12:52 PM on 5/16
          </span>
        </span>
      </span>
      <hr />
      <p
        class="PostDisplay-Content"
      >
         Expedita velit ea nesciunt odio. Illum dolorem dignissimos eos earum dolor eius iusto corporis. Neque inventore numquam vel.
      </p>
      <div>
        <span
          class="PostDisplay-ShowComments"
        >
          Show Comments
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PostDisplay should render PostDisplay snapshot 1`] = `<DocumentFragment />`;

exports[`PostDisplay should render PostDisplay with post 1`] = `
<DocumentFragment>
  <div
    class="sr-modal"
    style="z-index: 1;"
  >
    <div
      class="PostDisplay sr-modal_content"
      style="z-index: 1;"
    >
      <div
        class="split-between"
      />
      <div
        class="PostDisplayHeading"
        data-testid="PostDisplay-Heading"
      >
        <h4
          class="PostDisplay-Title"
        >
          Nulla saepe et nemo molestiae ducimus provident.
        </h4>
        <div
          class="PostDisplay-Buttons"
        >
          <button
            class="PinButton  PinButton IconButton  "
            data-testid="PinButton"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-thumbtack "
              data-icon="thumbtack"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 384 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M32 32C32 14.3 46.3 0 64 0H320c17.7 0 32 14.3 32 32s-14.3 32-32 32H290.5l11.4 148.2c36.7 19.9 65.7 53.2 79.5 94.7l1 3c3.3 9.8 1.6 20.5-4.4 28.8s-15.7 13.3-26 13.3H32c-10.3 0-19.9-4.9-26-13.3s-7.7-19.1-4.4-28.8l1-3c13.8-41.5 42.8-74.8 79.5-94.7L93.5 64H64C46.3 64 32 49.7 32 32zM160 384h64v96c0 17.7-14.3 32-32 32s-32-14.3-32-32V384z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
      <span
        class="PostedByOn"
      >
        <span>
           posted by 
        </span>
        <span>
          <a
            href="/users/user"
          >
            user
          </a>
        </span>
        <span
          class=" PostDisplay-TimeStamp"
        >
          <span
            class=""
          >
              at  12:52 PM on 5/16
          </span>
        </span>
      </span>
      <hr />
      <p
        class="PostDisplay-Content"
      >
         Expedita velit ea nesciunt odio. Illum dolorem dignissimos eos earum dolor eius iusto corporis. Neque inventore numquam vel.
      </p>
      <div>
        <span
          class="PostDisplay-ShowComments"
        >
          Show Comments
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
